src/pybind/mgr/mgr_util.py

```diff
@@ -106,3 +119,34 @@ def is_ipv6_enabled():
         get_default_addr.result = result
         return result
 
+
+def get_rates_from_data(data):
+    """
+    >>> get_rates_from_data([])
+    [(0, 0.0)]
+    >>> get_rates_from_data([[1, 42]])
+    [(1, 0.0)]
+    >>> get_rates_from_data([[0, 100], [2, 101], [3, 100], [4, 100]])
+    [(2, 0.5), (3, 1.0), (4, 0.0)]
+    """
+    if not data:
+        return [(0, 0.0)]
+    if len(data) == 1:
+        return [(data[0][0], 0.0)]
+    return [(data2[0], differentiate(data1, data2)) for data1, data2 in pairwise(data)]
+
+
+def differentiate(data1, data2):
```

Comment:
That's not true, if you reduce PGs of a pool for example the read / write ops and bytes decrease without any overflow - that was the reason to use the absolute value instead. (But this could also be a bug in ceph)

Reply:
I don't understand one thing, why can't we return a negative number here. If the derivative function returns a negative value it means that something is decreasing. I'm afraid by taking the absolute value we are returning incorrect data.

Reply:
One segment of a time series data often conists of the following:
`[tiemstamp, sum of something]` but this leaves the question open how should our rates look like?
In the end we want to show a graph that clearly shows if something is going on.

For example if we decrease the pgs of a pool with some data on it the time series sum for writing bytes and operation will drop over time.
As our writing speed and writing operations cannot grow negative we should show the values positive, as this would distract the user and would mess up the current graph.
To fix the graph we would have to center zero leaving us with half the space to display positive values for this rare case of negative values.

Nevertheless it could also be a bug in the time series usage for this particular case.

Reply:
> For example if we decrease the pgs of a pool with some data on it the time series sum for writing bytes and operation will drop over time.

Maybe this is the problem. Why would the value of written bytes decrease when the PG number decreases? Can you reproduce this behavior?

Nevertheless, we shouldn't use  `abs()` because from a mathematical perspective is just wrong.

Reply:
Yes I can reproduce this, @callithea wrote me a [comment](https://github.com/ceph/ceph/pull/28153#issuecomment-493469846) about this on an already merged PR where I already introduced the use of `abs`.

Mathematical it's presenting the absolute difference.
Using just the difference, would cause problems if the data we are receiving is not to 100% accurate. With decreasing sums and duplicate timestamps, we don't have that yet IMO.

In the end we don't want our users to freak out because their writing speed is negative ;), as it's like driving backwards with the car and it shows -10 km/h on the tachometer :smile:

Reply:
> Yes I can reproduce this, @callithea wrote me a [comment](https://github.com/ceph/ceph/pull/28153#issuecomment-493469846) about this on an already merged PR where I already introduced the use of `abs`.
> 
> Mathematical it's presenting the absolute difference.
> Using just the difference, would cause problems if the data we are receiving is not to 100% accurate. With decreasing sums and duplicate timestamps, we don't have that yet IMO.
> 
> In the end we don't want our users to freak out because their writing speed is negative ;), as it's like driving backwards with the car and it shows -10 km/h on the tachometer

That's why we should fix the root cause of the problem and not change the derivative function definition. We must understand why the total number of write/read bytes decreases when the PG number decreases.

Reply:
I'm with @rjfd here. If this all comes from a cumulative perf counter (read/written bytes to a pool), it should never decrease, unless that counter exhibits some weird semantics (on PG shrink written bytes are proportionally shrunk?). In any case that would smell like something's happening there :question: